gpu/matrix/coo_kernels.cu

```diff
@@ -0,0 +1,386 @@
+{
+    ValueType add_val;
+#pragma unroll
+    for (int i = 1; i < 32; i <<= 1) {
+        const IndexType add_ind = warp::shuffle_up(*ind, i);
+        add_val = zero<ValueType>();
+        if (threadIdx.x >= i && add_ind == *ind) {
+            add_val = *val;
+            if (i == 1) {
+                *head = false;
+            }
+        }
+        add_val = warp::shuffle_down(add_val, i);
+        if (threadIdx.x < 32 - i) {
+            *val += add_val;
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(128) void spmv_kernel(
+    const size_type nnz, const size_type num_lines,
+    const ValueType *__restrict__ val, const IndexType *__restrict__ col,
+    const IndexType *__restrict__ row, const ValueType *__restrict__ b,
+    ValueType *__restrict__ c)
+{
+    ValueType temp_val = zero<ValueType>();
+    const auto start = static_cast<size_type>(blockDim.x) * blockIdx.x *
+                           blockDim.y * num_lines +
+                       threadIdx.y * blockDim.x * num_lines;
+    size_type num = (nnz > start) * ceildiv(nnz - start, 32);
+    num = min(num, num_lines);
+    const IndexType ind_start = start + threadIdx.x;
+    const IndexType ind_end = ind_start + (num - 1) * 32;
+    IndexType ind = ind_start;
+    bool atomichead = true;
+    IndexType temp_row = (ind < nnz) ? row[ind] : 0;
+    IndexType next_row;
+    for (; ind < ind_end; ind += 32) {
+        temp_val += (ind < nnz) ? val[ind] * b[col[ind]] : zero<ValueType>();
+        next_row = (ind + 32 < nnz) ? row[ind + 32] : row[nnz - 1];
+        // segmented scan
+        const bool is_scan = temp_row != next_row;
+        if (warp::any(is_scan)) {
+            atomichead = true;
+            segment_scan(&temp_row, &temp_val, &atomichead);
+            if (atomichead) {
+                atomic_add(&(c[temp_row]), temp_val);
+            }
+            temp_val = 0;
+        }
+        temp_row = next_row;
+    }
+    if (num > 0) {
+        ind = ind_start + (num - 1) * 32;
+        temp_val += (ind < nnz) ? val[ind] * b[col[ind]] : zero<ValueType>();
+        // segmented scan
+        atomichead = true;
+        segment_scan(&temp_row, &temp_val, &atomichead);
+        if (atomichead) {
+            atomic_add(&(c[temp_row]), temp_val);
+        }
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_block_size) void set_zero(
+    const size_type nnz, ValueType *__restrict__ val)
+{
+    const auto ind =
+        static_cast<size_type>(blockDim.x) * blockIdx.x + threadIdx.x;
+    if (ind < nnz) {
+        val[ind] = zero<ValueType>();
+    }
+}
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void spmv(std::shared_ptr<const GpuExecutor> exec,
+          const matrix::Coo<ValueType, IndexType> *a,
+          const matrix::Dense<ValueType> *b, matrix::Dense<ValueType> *c)
+{
+    int multiple = 8;
```

Comment:
Hard-coding values like this is always difficult to warrant. I guess here it is probably okay. But maybe it should always accompany some metrics/ reasoning as to why this value was chosen.

Reply:
I agree with you that it shouldn't look like this.

However, AFAIK, these are tuned parameters obtained from running experiments (and might differ for different architectures). I also have a similar problem in #49, and have them hardcoded.
The problem is that we don't have this problem resolved in Ginkgo - the idea is to be able to say something like:  "this is a parameter that should be tuned, valid values are from the set __S__, and use __x__ by default". Then the user should be able to do an autotuning run of ginkgo, which will set the parameters to proper values for their system.

However we do it right now, there's a good chance we'll have to change it when we implement autotuning, so I'm fine for now with having it like this + adding a "TODO" that this should be changed when we support autotuning.